### ISSUE TITLE
Port twisted.internet.gireactor to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6055]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6055|
|Type|enhancement|
|Created|2012-10-04 18:01:11Z|
|Branch|https://github.com/twisted/twisted/tree/gireactor-py3-6055-4|

`twisted.internet.gireactor` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6055 6055
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_gireactor_py3_6055_4 branches/gireactor-py3-6055-4
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349373671000000 1349373671000000
changetime__1354648971000000 1354648971000000
version__None None
owner__itamar itamar
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35949]) Branching to 'gireactor-py3-6055'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [36256]) Branching to 'gireactor-py3-6055-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Sigh. This evolved into a combination of porting to Python 3, and updating the code to work with the versions of gtk and pygobject in Ubuntu 12.10. I will move latter into separate ticket once [#6161](https://github.com/twisted/twisted/issues/6161) is merged, except in the unlikely event you'd like to review the two combined.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [36430]) Branching to 'gireactor-py3-6055-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Merged forward, requires X to run tests (until [#6170](https://github.com/twisted/twisted/issues/6170) is fixed).

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gireactor-py3-6055-3

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Might it not be more idiomatic, in `gireactor.py`, to write `_pygtkcompatPresent = "gi.pygtkcompat" in theSystemPath`?
  1. Do end users really care about changed test behavior in obscure conditions?  Maybe they do...

http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/275/steps/shell_1/logs/stdio seems like a problem.  Perhaps we shouldn't add this test module to the list to run on Python 3 yet, so the rest can continue to run?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

1.I'll fix that, yeah.
 2. Oops, that's left over from when [[#6170](https://github.com/twisted/twisted/issues/6170)](https://github.com/twisted/twisted/issues/6170) was in this branched. Will switch to misc.
 3. That buildbot error is what #6170 fixes...

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Actually, can't use `in` - see [#6198](https://github.com/twisted/twisted/issues/6198).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [36516]) Branching to 'gireactor-py3-6055-4'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, ready for review again. Hopefully buildbot will be happy now.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gireactor-py3-6055-4

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @itamarst|
|-|-|

> Hopefully buildbot will be happy now.

eventually it decided to be...

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set status to `closed`|
|-|-|

(In [36519]) Merge gireactor-py3-6055-4.

Author: itamar
Review: exarkun
Fixes: [#6055](https://github.com/twisted/twisted/issues/6055)

Port gireactor to Python 3.
